Std/Tactic/Basic.lean

```diff
@@ -141,3 +141,16 @@ macro (name := triv) "triv" : tactic =>
 
 /-- `conv` tactic to close a goal using an equality theorem. -/
 macro (name := Conv.exact) "exact " t:term : conv => `(conv| tactic => exact $t)
+
+/-- The `conv` tactic `equals` claims that the currently focused subexpression is equal
+ to the given expression, and proves this claim using the given tactic.
+```
+example (P : (Nat → Nat) → Prop) : P (fun n ↦ n - n) := by
+  conv in (_ - _) => equals 0 by
+    -- current goal: ⊢ n - n = 0
+    apply Nat.sub_self
+  -- current goal: P (fun n => 0)
+```
+-/
+macro (name := Conv.equals) "equals " t:term " by " tac:tacticSeq : conv =>
+ `(conv|tactic => show (_ = $t); ($tac) )
```

Comment:
I think we want to be sure the tactic sequence closes the goal, otherwise it might be confusing to a user of this that in `conv`-mode they might get non-`conv` goals.
```suggestion
  `(conv| tactic => next => show (_ = $t); ($tac))
```

Regarding the syntax, using `by` like this is unique. Usually it's `=>` to indicate a tactic sequence (examples: `next => ...`, `case t => ...`, `induction` match arms, `conv => ...`, `focus => ...`, etc.), and there are some exceptions like `repeat` or `classical` that don't use an indicator at all. I think we should use `=>` for consistency.

In addition to this, we could have the following variant:
```lean
macro (name := Conv.equals') "equals " t:term " := " v:term : conv =>
  `(conv| tactic => exact ($v : _ = $t))
```
and you could write `equals 0 := by apply Nat.sub_self` if you'd rather see `by`. Having such a variant is consistent with tactics like `have`.

Reply:
I like that variant, even more composable. Will update the PR later (and also look for how to write a few small tests for this.)

Reply:
Hmm, that doesnt' work so nicely; it seems one then has to indent the proof a lot, like this:
```
example (P : (Nat → Nat) → Prop) : P (fun n ↦ n - n) := by
  conv in (_ - _) => equals 0 := by
                      -- current goal: ⊢ n - n = 0
                      apply Nat.sub
```
This makes logical sense, but isn't very helpful here.

Reply:
I think you might have to temper expectations about doing things on one line here. At least there's
```lean
example (P : (Nat → Nat) → Prop) : P (fun n ↦ n - n) := by
  conv in (_ - _) =>
    equals 0 := by
      apply Nat.sub
```
or
```
example (P : (Nat → Nat) → Prop) : P (fun n ↦ n - n) := by
  conv in (_ - _) =>
  equals 0 := by
  apply Nat.sub
```

Reply:
Maybe … but with `" => " tacticSeq` as it stands right now, the indentation is more flexible, so I might have a slight preference for this variant after all (and two character less to type)